Use Internal Clock for effects when no beatgrid is available

kshitij98 opened the pull request:
https://bugs.launchpad.net/mixxx/+bug/1738590

I have tried to fix this by using the `bpm` and `beat_distance` present in `[InternalClock]` while collecting group features when no beatgrid is available.

kshitij98 committed:
Use Internal Clock when no beatgrid is available

daschuer commented:
Thank you for this PR, I think it works only partly. This is probably an issue of the [InternalClock] implementation. It is updaten when sync is activated. But I think it should work without actually pressing sync. It should somehow predict the bpm. 
The main use case is IMHO to sync the master effect. This should adopt the bpm of the "playing" deck for example.

kshitij98 commented:
Thank you for reviewing!
Okay, so using the bpm of the first playing deck with bpm info and resorting to internal clock if no such deck is available must handle all the use cases.

daschuer commented:
Yes, I assume that the logic is somewhere already. I am unsure what the purpose of the internal clock was originally, but I think we can use it and just update it more often.

daschuer commented:
Maybe @ywwg the original developer can give some hints.

Be-ing commented:
Thank you for working on this, but from a quick look at the code this seems to only partially solve the problem. As far as I can tell these changes will only affect decks and samplers that do not have a beatgrid. In addition to this, the master sync tempo should be passed via the GroupFeatures to the effects processing for the master mix and other mixed buses in EngineMaster::process.

kshitij98 committed:
Sync master and all active channels with Internal Clock

kshitij98 commented:
Yes, those changes only affected the decks and samplers without the beatgrid.

I have now updated the features of all active channels and master handle in the last commit. It currently syncs them with the internal clock if no beatgrid is available.

Be-ing commented on src/engine/enginemaster.cpp:
```diff
@@ -188,6 +188,9 @@ EngineMaster::EngineMaster(UserSettingsPointer pConfig,
     m_pHeadphoneEnabled->setReadOnly();
 
+    m_masterBpm = new ControlProxy("[InternalClock]", "bpm", this);
+    m_masterBeatDistance = new ControlProxy("[InternalClock]", "beat_distance", this);
```
This feels somewhat hacky to use ControlProxies for this. Couldn't this information be obtained from m_pMasterSync?

kshitij98 replied:
Updated. For this I had to make the `BaseSyncableListener::masterBpm()` and `BaseSyncableListener::masterBeatDistance()` functions public from protected.

Be-ing commented:
> Thank you for this PR, I think it works only partly. This is probably an issue of the [InternalClock] implementation. It is updaten when sync is activated. But I think it should work without actually pressing sync. It should somehow predict the bpm.
The main use case is IMHO to sync the master effect. This should adopt the bpm of the "playing" deck for example.

I verified this by running Mixxx with the `--developer` command line option and watching the value of `[InternalClock], bpm`. That will need to be updated whenever the master sync BPM is changed to make this work correctly.

Be-ing commented on src/engine/enginemaster.cpp:
```diff
@@ -356,6 +359,7 @@ void EngineMaster::processChannels(int iBufferSize) {
         if (m_pEngineEffectsManager) {
             GroupFeatureState features;
             pChannel->collectFeatures(&features);
+            collectMasterSyncFeatures(&features);
```
At first I thought this would always overwrite the BPM from `pChannel->collectFeatures` in the line above, but then I saw that `EngineMaster::collectMasterSyncFeatures` checks `!features->has_beat_length_sec`. I think it would help to add a comment here clarifying that the master BPM is only used if `pChannel->collectFeatures` does not report a BPM.

kshitij98 replied:
Updated.

Be-ing commented:
Use `EngineMaster::collectMasterSyncFeatures` for every call to `m_pEngineEffectsManager->processPostFaderInPlace` in `EngineMaster::process`.

ywwg commented:
The purpose of the Internal Clock is to avoid the need for a "master deck" like traktor requires.  When master sync is on, all decks are technically slaves to the internal clock.  However *any* deck can adjust the internal clock by adjusting its own pitch slider.  So in general the theory behind this PR is correct: using the internal clock to sync effects should work as a backup when a beatgrid is missing

ywwg commented:
I think having the Internal Clock be up to date when sync is off is possible, and ok, but when two decks are playing, and they both have sync off, which one should the internal clock be aligned to?  If I engage an effect on deck 1, I want it to be synced to deck 1, and if I engage an effect on deck 2, it should be synced to deck 2.  Those two decks may have very different bpms if the DJ is doing something funky

ywwg commented:
In the sync implementation, the internal clock is reset when a single deck with sync enabled starts playing and no other sync-enabled decks are playing.  Take a look at the exhaustive unit tests for an idea how this works

ywwg commented:
I think the correct solution is to create a mixer blinkenlight for the internal clock with a tap-tempo.  This is how hardware mixers work -- you tap the tempo you want to sync and go by that, and even if it's a little off the effects sound fine.  We probably should be showing the internal clock to the user anyway

ywwg commented:
There will be lots of UX questions like, does SYNC always override the tap tempo or now do we have to have yet another way to explicitly trigger that... it's not simple

Be-ing commented:
> when two decks are playing, and they both have sync off, which one should the internal clock be aligned to

I think it would be reasonable to decide this based on which deck had its play button pressed first. Should it be the deck whose play button was pressed first or pressed last?

> If I engage an effect on deck 1, I want it to be synced to deck 1, and if I engage an effect on deck 2, it should be synced to deck 2. Those two decks may have very different bpms if the DJ is doing something funky

Effect syncing already works when effects are applied to decks. This PR is dealing with the case of using the non-deck effect inputs like the master mix.

>  think the correct solution is to create a mixer blinkenlight for the internal clock with a tap-tempo. This is how hardware mixers work -- you tap the tempo you want to sync and go by that, and even if it's a little off the effects sound fine. We probably should be showing the internal clock to the user anyway ... There will be lots of UX questions like, does SYNC always override the tap tempo or now do we have to have yet another way to explicitly trigger that... it's not simple

Interesting ideas, but those are beyond the scope of this PR.

ywwg commented:
I think it'd be ok to tweak the master clock code to reset the following way:

* if the deck is the first playing deck with sync ON, reset the clock (regardless of any other playing decks) (this is the current behavior)
* if the deck is the first playing deck, period, reset the clock (new behavior)

We definitely still need to reset the clock when a SYNC deck starts playing or master sync won't work.  Please write new unit tests for the behavior you write, and I'd like to review them to make sure everything still makes sense.

ywwg commented:
hm, since the first playing deck will not have SYNC on, the Internal Clock may drift (esp if the DJ tweaks that deck).  So it's still not guaranteed that the internal clock will line up with any deck.  Furthermore if the DJ keeps continuously mixing, without ever using SYNC (beatmatching by ear) then the Internal Clock could be very much wrong after just a few transitions.  

Maybe instead of resetting the clock when the DJ his play, it needs to reset when the internal state changes to "only one deck is playing"?

kshitij98 committed:
Use m_pMasterSync instead of ControlProxy

kshitij98 committed:
Collecting Master Sync features for Headphones and Bus

Be-ing commented:
The new commits look good so far. Thank you for splitting the changes into small commits. Now the harder part is keeping the internal clock updated as the synced decks change.

kshitij98 committed:
Update internal clock without synced decks

kshitij98 committed:
Added setPlayingSyncable and setPausedSyncable functions

kshitij98 commented:
Thanks a lot @ywwg and @Be-ing for reviewing!

I have added a few commits. So the new behaviour I've implemented is:

- If no sync deck currently exists, set the internal clock bpm features to the earliest playing deck's.
For maintaining the time of the playing decks, I've implemented two functions `BaseSyncableListener::setPlayingSyncable(Syncable*)` and `BaseSyncableListener::setPausedSyncable(Syncable*)` which reorder the `m_syncables` list to keep the playing decks in the beginning with the first playing deck in the beginning of those playing decks.

A few things which can be improved imo are:
- The phase is not being reset on scratching and seeking. For this i tried logging `EngineSync::notifyScratching()` and `EngineSync::notifyBeatDistanceChanged()`, but these are not being triggered on scratching and seeking.

Also, By looking at the code I assumed that changing the ordering of `m_syncables` must not break any other feature.

Should I proceed on writing unit tests for this?

daschuer commented:
This may be related: 
https://mixxx.zulipchat.com/#narrow/stream/109695-support/topic/Sync.20issue.20(Master.20BPM.3F.3F)

kshitij98 commented:
Thank you for the reference.

Also, I forgot to mention:

- It only fetches the BPM info of first playing tracks if there are no sync enabled tracks. Even the presence of paused sync enabled tracks sync the master BPM to them.

Be-ing reviewed (changes requested):
The code looks quite good and worked well in my brief testing. It would help to add a bit more explanation of why the code is the way it is in the comments.

The changes required to implement this ended up being more extensive than I thought. I would be more comfortable merging this to master instead of 2.1 at this point.

Be-ing commented:
When I change `[InternalClock], bpm` without any synced decks, then press sync on the deck that I started playing second so its rate slider is changed to match the first playing deck, `[InternalClock], bpm` does not update to match the synced deck until the tempo of the synced deck is changed.

Be-ing commented:
> Should I proceed on writing unit tests for this?

Yes please.

kshitij98 committed:
Updated comments and conformance to coding standards

kshitij98 commented on src/engine/sync/basesyncablelistener.cpp:
```diff
@@ -184,3 +184,38 @@ void BaseSyncableListener::checkUniquePlayingSyncable() {
+            m_syncables[i] = m_syncables[i+1];
+            m_syncables[i+1] = temporarySyncable;
+        }
+        else if (!temporarySyncable->isPlaying()) {
```
Updated.

kshitij98 commented on src/engine/sync/basesyncablelistener.cpp:
```diff
@@ -184,3 +184,38 @@ void BaseSyncableListener::checkUniquePlayingSyncable() {
+void BaseSyncableListener::setPlayingSyncable(Syncable* pSource) {
+    Syncable *currentSyncable = pSource;
+    // Rearrange the syncables with the playing syncables clustered in the beginning
+    // and appending the current syncable in the end of the playing syncables
```
Updated.

kshitij98 commented on src/engine/sync/basesyncablelistener.cpp:
```diff
@@ -184,3 +184,38 @@ void BaseSyncableListener::checkUniquePlayingSyncable() {
+}
+
+void BaseSyncableListener::setPausedSyncable(Syncable* pSource) {
+    // Maintain the playing syncables clustering in the beginning
```
I have updated it to "Push the current syncable out the playing syncables group so that the playing syncables are at the beginning of the list".

Be-ing commented on src/engine/sync/basesyncablelistener.cpp:
```diff
@@ -187,9 +187,9 @@ void BaseSyncableListener::checkUniquePlayingSyncable() {
-    // and appending the current syncable in the end of the playing syncables
-    for (int i=0 ; i<m_syncables.length() ; ++i) {
+    // Sort the syncables with the playing syncables clustered in the beginning
+    // and appending the current syncable at the end of the playing syncables
```
Please add a note about the reason for the function here as well: "This allows for keeping track of the first played deck."

kshitij98 replied:
Updated.

kshitij98 committed:
Updated comments

kshitij98 commented:
> When I change [InternalClock], bpm without any synced decks, then press sync on the deck that I started playing second so its rate slider is changed to match the first playing deck, [InternalClock], bpm does not update to match the synced deck until the tempo of the synced deck is changed.

This feature was added explicitly previously:
```
            // There is no master. If any other deck is playing we will match
            // the first available bpm -- sync won't be enabled on these decks,
            // otherwise there would have been a master.
```
/src/engine/sync/enginesync.cpp : Line 103

Should I set the newly synced deck's bpm as master in this case?

Be-ing commented:
> Should I set the newly synced deck's bpm as master in this case?

I think so unless @ywwg suggests otherwise.

daschuer commented on src/engine/sync/basesyncablelistener.cpp:
```diff
@@ -184,3 +184,40 @@ void BaseSyncableListener::checkUniquePlayingSyncable() {
+    Syncable *currentSyncable = pSource;
+    // Sort the syncables with the playing syncables clustered in the beginning
+    // and appending the current syncable at the end of the playing syncables.
+    // This allows for keeping track of the first played deck, which is needed 
```
Here we should choose the current playing deck. 
It can be queried via PlayerInfo::getCurrentPlayingDeck()

kshitij98 replied:
This is an interesting idea. Using this, We won't need to keep track of the time of playing of decks. And it makes more sense to set the most audible deck as master if sync is turned off.

kshitij98 replied:
To implement this we'll have to create more listeners in EngineSync i guess. Otherwise it won't be able to update the master bpm on volume change.

daschuer replied:
The PlayerInfo class does this already.
Not sure if it meets you timing requirements though.
I would prefer to keep the current track logic in a single place.

kshitij98 replied:
Yes, It does not meet the timing requirements currently. Will need to implement more listeners for volume change & crossfader updates.

> I would prefer to keep the current track logic in a single place.

Can you elaborate?

daschuer replied:
Currently the PlayerInfo Class is resposible to decide which is the "current deck" and "current track". 
It would be nice to re-use it here. If it is not possible for some reasons it is also OK.

kshitij98 replied:
I was trying to add slots for volume level updates. But for some reason, they are not getting connected. `valueChanged()` signal is not emitted although the key "volume" is getting updated for that group. I checked this in the developer tools.

Can you review? I have pushed a commit for the same.

daschuer commented on src/engine/sync/basesyncablelistener.cpp:
```diff
@@ -184,3 +184,40 @@ void BaseSyncableListener::checkUniquePlayingSyncable() {
+    for (int i = 0; i < m_syncables.length(); ++i) {
+        Syncable* temporarySyncable = m_syncables[i];
+        if (temporarySyncable != pSource && temporarySyncable->isPlaying()) {
+            continue;
```
A source code comment would be nice here.

daschuer commented on src/engine/sync/basesyncablelistener.cpp:
```diff
@@ -184,3 +184,40 @@ void BaseSyncableListener::checkUniquePlayingSyncable() {
+        m_syncables[i] = currentSyncable;
+        currentSyncable = temporarySyncable;
+        if (temporarySyncable == pSource) {
+            break;
```
Why can we break here early? Please comment.

kshitij98 committed:
[WIP] Added volume level slot
